Generation/src/components/HepMCToEDMConverter.cpp

```diff
@@ -29,43 +32,79 @@ StatusCode HepMCToEDMConverter::execute() {
       HepMC::Units::conversion_factor(event->length_unit(), gen::hepmcdefault::length) * gen::hepmc2edm::length;
   double hepmc2EdmEnergy =
       HepMC::Units::conversion_factor(event->momentum_unit(), gen::hepmcdefault::energy) * gen::hepmc2edm::energy;
+  
+  // bookkeeping of particle / vertex relations
+  std::unordered_map<const HepMC::GenVertex*, fcc::GenVertex> hepmcToEdmVertexMap;
+  HepMC::FourVector tmp; /// temp variable for the transfer of position / momentom
+  // iterate over particles in event
+  for (auto particle_i = event->particles_begin(); particle_i != event->particles_end(); ++particle_i) {
 
-  // currently only final state particles converted (no EndVertex as they didn't decay)
-  // TODO add translation of decayed particles
-  HepMC::FourVector tmp;
-  for (auto vertex_i = event->vertices_begin(); vertex_i != event->vertices_end(); ++vertex_i) {
-    tmp = (*vertex_i)->position();
-    auto vertex = vertices->create();
-    auto& position = vertex.position();
-    position.x = tmp.x() * hepmc2EdmLength;
-    position.y = tmp.y() * hepmc2EdmLength;
-    position.z = tmp.z() * hepmc2EdmLength;
-    vertex.ctau(tmp.t() * Gaudi::Units::c_light * hepmc2EdmLength);  // is ctau like this?
+    // if there is a list of statuses to filter: filter by status
+    if (m_hepmcStatusList.size() > 0) {
+      if(std::find(m_hepmcStatusList.begin(), m_hepmcStatusList.end(), (*particle_i)->status()) != m_hepmcStatusList.end()) continue;
+    }
+    // create edm 
+    fcc::MCParticle particle = particles->create();
+    // set mcparticle data members
+    particle.pdgId((*particle_i)->pdg_id());
+    particle.status((*particle_i)->status());
+    /// lookup charge in particle properties
+    HepPDT::ParticleID particleID((*particle_i)->pdg_id());
+    particle.charge(particleID.charge());
+
+    auto& p4 = particle.p4();
+    tmp = (*particle_i)->momentum();
+    p4.px = tmp.px() * hepmc2EdmEnergy;
+    p4.py = tmp.py() * hepmc2EdmEnergy;
+    p4.pz = tmp.pz() * hepmc2EdmEnergy;
+    p4.mass = (*particle_i)->generated_mass() * hepmc2EdmEnergy;
 
-    for (auto particle_i = (*vertex_i)->particles_begin(HepMC::children);
-         particle_i != (*vertex_i)->particles_end(HepMC::children);
-         ++particle_i) {
-      // take only final state particles
-      if ((*particle_i)->status() != 1) continue;
+    /// create production vertex, if it has not already been created and logged in the map
+    HepMC::GenVertex* productionVertex = (*particle_i)->production_vertex();
+    if (nullptr != productionVertex) {
+      if (hepmcToEdmVertexMap.find(productionVertex) != hepmcToEdmVertexMap.end()) {
+        // vertex already in map, no need to create a new one
+        particle.startVertex(hepmcToEdmVertexMap[productionVertex]);
```

Comment:
please change: from "hepmcToEdmVertexMap[productionVertex]" to "hepmcToEdmVertexMap[decayVertex]"

Reply:
Hi Loukas @gouskos,

thanks for your comments and sorry for the late reply! The way you write it it doesn't work - `decayVertex` isn't declared until line 83 below. There was a copy & paste bug introduced by me, which I already fixed in https://github.com/HEP-FCC/FCCSW/pull/243/commits/76c9252a2fc332d818b4fdfa2ced8e924489b446
it's essentially the same thing but on L87. Could I ask you to check that this doesn't work already? I'll have a look why the tests are failing!
Cheers,
Valentin